### ISSUE TITLE
reporting unused private vars inside macros

### ISSUE THREAD

**Issue by barkanido** [borkdude/clj-kondo#590]
**version**

v2019.11.03

**platform**

OsX, Clojure 1.10, Java "1.8.0_162"

**problem**

clj-kondo reports `Unused private var` for vars inside `mount.core/defstate`

**repro**

```
(defn- init-db [] ...)

(mount.core/defstate clicks-db
                     :start (init-db)
```

**expected behavior**

clj-kondo should understand that `init-db` is _used_ somehow, or at least silence this warning  for specific macros (like `mount.core/defstate`)

**actual behavior**

`warning: Unused private var some-ns/init-db`

**Comment by borkdude** [comment]
@barkanido Can you please provide a minimal but complete lintable example, including a namespace declaration, so I can paste the code in a `repro.clj` file and can run:

```
clj-kondo --lint repro.clj --config '...'
```

where `...` is the configuration you are currently using for mount.core, if any.

**Comment by barkanido** [comment]
sure:

```
$cat src/af_matcher_ng/repository/aerospike.clj
(ns af-matcher-ng.repository.aerospike
    (:require [taoensso.timbre :refer [warn error info spy]]
              [mount.core :refer [defstate]])
  (:import [com.aerospike.client.async NioEventLoops]))

(defn- init-aerospike [] 0)

(defstate clicks-db
          :start (assoc (init-aerospike) :foo "bar"))
```

```
$ cat .clj-kondo/config.edn
{:linters
 {:unresolved-symbol {:exclude [(mount.core/defstate)
                                (clojure.core.match/match)
                                (compojure.core/routes)
                                (clojure.java.jdbc/with-db-transaction)]
                      :level :info}
  :unused-referred-var {:level :off
                        :exclude {taoensso.timbre [info spy debug error warn]}}
  :unused-binding {:exclude-destructured-keys-in-fn-args true
                   :level :info}
  :unused-namespace {:exclude [taoensso.timbre]
                     :level :off}}
 :lint-as {mount.core/defstate clojure.core/declare}
 :consistent-alias {:aliases {clojure.string s}}}
```

```
$ clj-kondo --lint src/af_matcher_ng/repository/aerospike.clj --cache false
src/af_matcher_ng/repository/aerospike.clj:6:1: warning: Unused private var af-matcher-ng.repository.aerospike/init-aerospike
linting took 11ms, errors: 0, warnings: 1
```
I have resolved the `unresolved symbol` by using `:lint-as {mount.core/defstate clojure.core/declare}`.

**Comment by miikka** [comment]
For what is worth, I'm using `:lint-as {mount.core/defstate clojure.core/def}` and it resolves this problem.

**Comment by borkdude** [comment]
@barkanido I suggest you use @miikka's config. I'll try to include it as a built-in config for a next release.

**Comment by barkanido** [comment]
@borkdude thanks for the swift response!
using `:lint-as {mount.core/defstate clojure.core/def}` instead of `:lint-as {mount.core/defstate clojure.core/declare}` solves this.

Btw, `:lint-as {mount.core/defstate clojure.core/declare}` is an idea from the docs so it might be better to update them.

**Comment by borkdude** [comment]
@barkanido Can you point to the exact location in the docs where you picked this up?

**Comment by barkanido** [comment]
https://github.com/borkdude/clj-kondo/blob/master/doc/config.md#exclude-unresolved-symbols-from-being-reported

> Sometimes vars are introduced by executing macros, e.g. when using HugSQL's def-db-fns. You can suppress warnings about these vars by using declare. Example:...

**Comment by borkdude** [comment]
@barkanido I wasn't suggesting to use `:lint-as declare` there, but to literally write:
`(declare foo)` to suppress warnings about an unknown var `foo`.
